### ISSUE TITLE
MergeLayer accepting dictionaries

### ISSUE THREAD

**Issue by f0k** [Lasagne/Lasagne#537]
Right now, our MergeLayer base class, the only layer class that can handle multiple inputs, requires its input layers to be given as a list. This is a good fit for the ConcatLayer, for example, which concatenates multiple input tensors in a specific order. In #522, we found that this is a bad fit for the recurrent layers, though, which accept multiple input layers meaning very different things: the actual input, a mask, and an initial hidden state. For this use case, it would be much more suitable to access the different input layers (and inputs, and input shapes) by name: that is, `incomings` in the constructor, `input_shapes` in `get_output_shape_for()` and `inputs` in `get_output_for()` should be dictionaries rather than lists.

What can we do to allow this?
a) Add another `MergeLayer` base class (NamedMergeLayer` or `DictMergeLayer` or something) that has a different interface and adapt `lasagne.layers.get_output()` accordingly?
b) Expand `MergeLayer` to accept either a list or a dictionary in its constructor and change its behavior accordingly? -> This would mean that existing merge layers need to cope with the possibility of being handed a dictionary.
c) ~~Change the `Layer` base class to always allow multiple input layers?~~ -> This wouldn't make sense for most layers such as the `DenseLayer`.

There's also the side issue of #337: We may want to allow a single layer to output a list or dictionary of tensors. How would that interplay with merge layers? Is is to be treated completely orthogonal (i.e., a merge layer is for merging multiple layers, not for processing multiple inputs) or do we need to account for this?

At first glance, a) seems like the cleanest solution, with the disadvantage that having two different merge layer base classes will be confusing, especially if we cannot rename the current one.

**Comment by craffel** [comment]
> a) Add another MergeLayer base class (NamedMergeLayer or DictMergeLayer or something) that has a different interface and adapt lasagne.layers.get_output() accordingly?

I think the current `MergeLayer` has semantics which are different enough (expects a list of inputs which can be concatenated, where order matters) from a "named" merge layer (accepts certain inputs) that having them separate is ok with me.  If we change `MergeLayer` to accept dicts, that would also maybe break user-created `MergeLayer` subclasses (if a dict was passed).  So I'm in support of a).

**Comment by skaae** [comment]
> There's also the side issue of #337: We may want to allow a single layer to output a list or dictionary of tensors. How would that interplay with merge layers? Is is to be treated completely orthogonal (i.e., a merge layer is for merging multiple layers, not for processing multiple inputs) or do we need to account for this?

I think it would be a nice addition if layers could return several outputs. Two use cases are a) encoders/decoders where you want to return both the cell and the hidden state and b) variational methods where you need to access the statistics for the latent variables.

**Comment by benanne** [comment]
I like @f0k's idea of allowing this `DictMergeLayer` to work with both separate layers providing the different inputs, and a single layer providing a dict of inputs (or combinations of the two). Although I'm having a hard time imagining the API for this :) We should definitely investigate this though. Seems like the most flexible solution.

Maybe we could also generalize this to the ordered case (i.e. the current `MergeLayer`), where a layer could provide a list of outputs and `MergeLayer` subclasses would do the right thing. Although I guess that has fewer practical applications.

**Comment by f0k** [comment]
> I like @f0k's idea of allowing this DictMergeLayer to work with both separate layers providing the different inputs, and a single layer providing a dict of inputs (or combinations of the two). Although I'm having a hard time imagining the API for this :)

Yes, exactly. It might be easier to treat this independently -- i.e., merge layers are strictly for merging multiple layers, and *every* layer may choose to return an output dictionary/list, or choose to accept an input dictionary/list from its input layer(s). IIRC, the latter wouldn't even require any changes to `get_output()`. I'll post some thoughts on this in #337, though.
The alternative would be that all layers are allowed to produce multiple outputs, but only MergeLayers are allowed to consume multiple inputs. This would enforce that every layer accepting multiple inputs could be fed either from a single multi-output layer or multiple single-output layers. That might be good as well... hmm... it turns out that the two issues are quite entangled!

> If we change `MergeLayer` to accept dicts, that would also maybe break user-created `MergeLayer` subclasses (if a dict was passed).

Yes, when forcefully merging both semantics into one base class, we'd end up with subclasses that only support one behavior or the other. Probably there aren't any use cases that would require both ways to be accepted, so it's better to split them right away.
This still leads us into another pesky name discussion. We cannot change the name of `MergeLayer` for compatibility purposes, but we could add a new alias for it, or rename it and alias the new name to `MergeLayer`.
What would be a good pair of names then? `DictMergeLayer` and `ListMergeLayer`? `NamedMergeLayer` and `OrderedMergeLayer`?

**Comment by mjwillson** [comment]
While you're discussing this, is there a reason why MergeLayer needs to be a special subclass, could support for multiple incomings be rolled into the Layer superclass, and a single incoming just be a special case of it?

Currently MergeLayer inherits from Layer but actually has a different, partly incompatible interface, which leads to some ugly isinstance checks in a few places. Also means things like ExpressionLayer don't work with multiple inputs, unless you copy-paste another parallel subclass ExpressionMergeLayer.

If anything it feels to me like Layer (accepts one input) is a special case of MergeLayer (accepts any number of inputs), rather than vice-versa.

See also InputLayer, which is another special subclass of Layer with IIRC an incompatible interface that requires isinstance checks in some places. It's a layer which has zero incomings, perhaps could be treated just as another special case of "any number of incomings"?

**Comment by f0k** [comment]
> While you're discussing this, is there a reason why MergeLayer needs to be a special subclass, could support for multiple incomings be rolled into the Layer superclass, and a single incoming just be a special case of it?

I think the main reason is what we called "Restraint" (for the lack of a better word) in our [design principles](http://lasagne.readthedocs.org/en/latest/user/development.html#philosophy). The most common type of layer only has a single input, and a single incoming layer. It should be straightforward to read and write such implementations, without having to know or care that there could be other cases. I.e., when implementing a dense layer, users should not have to check that they receive a single input only, or are linked to a single incoming layer only, or have to unwrap the incoming data in some way.
But you're right that inheritance could probably be turned around such that `Layer` is a special case that has a single incoming layer, and `InputLayer` is a special case that has no incoming layers (although we still need to differentiate that from the case of a free-floating layer such as `DenseLayer((10, 20), 30)`). However, this would require additional wrapper methods such that implementers of a `Layer` subclass can access `self.input_layer` instead of `self.input_layers[0]` and their `get_output_for(inputs, **kwargs)` does not have to unwrap `inputs[0]`. This might be more difficult to follow than what we have now (although I'm also unhappy that `InputLayer` and `MergeLayer` break the `Layer` interface).
A further complication is that Lasagne is already in use, and changing old design decisions is difficult without breaking code. Maybe we should collect ideas for a major (non-compatible) release somewhere.

Is there a way to fix things in a straightforward and backwards-compatible way? Maybe add a new `BaseLayer` base class that allows `incoming` to be any collection of layers/shapes (list, dictionary, or plain) and have `self.input_shape` and `input` in `get_output_for(input)` match this structure? Then have `Layer` ensure that `incoming` is a plain layer or shape, `MergeLayer` ensure that it is a plain list, and `InputLayer` ensure that it is a plain shape? We'd still need to have MergeLayer and InputLayer inherit from Layer (which is ugly) to ensure backwards compatibility (people expecting `isinstance(x, Layer)` to catch all kinds of layers), but they could just override whatever Layer does.
